packages/angular/cli/lib/examples/signal-forms/02-built-in-validators.md

```diff
@@ -0,0 +1,209 @@
+  max(form.age, 100);
+
+  required(form.email);
+  email(form.email);
+
+  pattern(form.website, /^(https?:\/\/)?([\da-z\.-]+)\.([a-z\.]{2,6})([\/\w \.-]*)*\/?$/);
+});
+
+@Component({
+  selector: 'app-user-settings',
+  imports: [JsonPipe, Control],
+  templateUrl: './user-settings.component.html',
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class UserSettingsComponent {
+  readonly submitted = output<UserSettings>();
+
+  settingsModel = signal<UserSettings>({
+    username: '',
+    age: 18,
+    email: '',
+    website: '',
+  });
+
+  settingsForm = form(this.settingsModel, settingsSchema);
+
+  async handleSubmit() {
+    await submit(this.settingsForm, async () => {
+      this.submitted.emit(this.settingsForm().value());
+    });
+  }
+}
+```
+
+### user-settings.component.html
+
+This file provides the template for the form, displaying specific error messages for each potential validation failure.
+
+```html
+<form (submit)="handleSubmit(); $event.preventDefault()">
+  <div>
+    <label>Username:</label>
+    <input type="text" [control]="settingsForm.username" />
+    @if (settingsForm.username().errors().length > 0) {
+    <div class="errors">
+      @for (error of settingsForm.username().errors(); track error) { @switch (error.kind) { @case
+      ('required') {
+      <p>Username is required.</p>
+      } @case ('minLength') {
+      <p>Username must be at least 3 characters.</p>
+      } @case ('maxLength') {
+      <p>Username cannot exceed 20 characters.</p>
+      } } }
+    </div>
+    }
+  </div>
+
+  <div>
+    <label>Age:</label>
+    <input type="number" [control]="settingsForm.age" />
+    @if (settingsForm.age().errors().length > 0) {
+    <div class="errors">
+      @for (error of settingsForm.age().errors(); track error) { @switch (error.kind) { @case
+      ('required') {
+      <p>Age is required.</p>
+      } @case ('min') {
+      <p>You must be at least 18 years old.</p>
+      } @case ('max') {
+      <p>Age cannot be more than 100.</p>
+      } } }
+    </div>
+    }
+  </div>
+
+  <div>
+    <label>Email:</label>
+    <input type="email" [control]="settingsForm.email" />
+    @if (settingsForm.email().errors().length > 0) {
+    <div class="errors">
+      @for (error of settingsForm.email().errors(); track error) { @switch (error.kind) { @case
+      ('required') {
+      <p>Email is required.</p>
+      } @case ('email') {
+      <p>Please enter a valid email address.</p>
+      } } }
+    </div>
+    }
+  </div>
+
+  <div>
+    <label>Website:</label>
+    <input type="url" [control]="settingsForm.website" />
+    @if (settingsForm.website().errors().length > 0) {
+    <div class="errors">
+      @for (error of settingsForm.website().errors(); track error) { @switch (error.kind) { @case
+      ('pattern') {
+      <p>Please enter a valid URL.</p>
+      } } }
+    </div>
+    }
+  </div>
+
+  <button type="submit" [disabled]="!settingsForm().valid()">Save Settings</button>
```

Comment:
It is considered a bad pattern to disable buttons. It makes forms less accessible.
Plus, if the only way to submit the form is by clicking the button it's also not good. Entering for example will also submit the form if the submit handler on the form element is implemented.

```suggestion
  <button type="submit">Save Settings</button>
```